### PR TITLE
Read multi indexes

### LINKED ISSUE #262: ENH: support for reading MultiIndex
From discussion in https://github.com/pandas-dev/pandas/pull/18629 (writing and reading indexes in `to_parquet` / `read_parquet` the default behaviour).

`fastparquet` supports writing multiple index levels and the pandas metadata are written correctly (and `pyarrow` can also read them in correctly), but raises an NotImplementedError on reading them in:

```
In [1]: import fastparquet

In [2]: fastparquet.__version__
Out[2]: '0.1.3'

In [3]: df = pd.DataFrame(np.random.randn(10, 3), columns=list('abc'), index=pd.MultiIndex.from_product([['A', 'B'], range(5)]))

In [5]: fastparquet.write("__test_mi.parquet", df)

In [6]: f = fastparquet.ParquetFile("__test_mi.parquet")

In [8]: f.key_value_metadata['pandas']
Out[8]: '{"columns": [{"metadata": null, "name": "level_0", "numpy_type": "object", "pandas_type": "unicode"}, {"metadata": null, "name": "level_1", "numpy_type": "int64", "pandas_type": "int64"}, {"metadata": null, "name": "a", "numpy_type": "float64", "pandas_type": "float64"}, {"metadata": null, "name": "b", "numpy_type": "float64", "pandas_type": "float64"}, {"metadata": null, "name": "c", "numpy_type": "float64", "pandas_type": "float64"}], "index_columns": ["level_0", "level_1"], "pandas_version": "0.21.0"}'

In [9]: f.to_pandas()
...
NotImplementedError: multi-index not yet supported, use index=False
```

(if you don't like opening issues for known limitations, it's fine to close)

### PR DESCRIPTION
**Pull request by 0x0L**
Closes #262

to_pandas() can accept a list of columns

For categorical multi-index it's a bit trickier. I'm not too sure what to do (subclass MultiIndex?) to make it work. Help would be greatly appreciated :)

### CODE REVIEW

**Comment by martindurant** [comment]
@TomAugspurger , any opinion on how to make a multi-index out of categorical columns?

**Comment by TomAugspurger** [comment]
Sorry, what's the question? A MultiIndex with multiple categorical levels?

**Comment by martindurant** [comment]
A multi-index, where one or more of the levels should be categorical. They get made in to a list of indexes [here](https://github.com/dask/fastparquet/pull/331/files#diff-595248c4beed78802cbcd980a66eebfcR69) and then passed to the MultiIndex constructor - but seem to loose their category nature along the way.
 
@0x0L , did I understand the situation?

**Comment by 0x0L** [comment]
@martindurant 
Yes that's it.

**Comment by TomAugspurger** [comment]
@0x0L can you point to the line where the categorical is lost?

FYI, the `test_multi_index_category` test fails on pandas master with

```
            if not level.is_unique:
                raise ValueError("Level values must be unique: {values} on "
                                 "level {level}".format(
                                     values=[value for value in level],
>                                    level=i))
E               ValueError: Level values must be unique: [Timestamp('1970-01-01 00:00:04.322432592'), Timestamp('1970-01-01 00:00:04.322432592'), Timestamp('1970-01-01 00:00:04.321907912')] on level 0

../pandas/pandas/core/indexes/multi.py:283: ValueError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/taugspurger/sandbox/pandas/pandas/core/indexes/multi.py(283)_verify_integrity()
-> level=i))
(Pdb) up
> /Users/taugspurger/sandbox/pandas/pandas/core/indexes/multi.py(240)__new__()
-> result._verify_integrity()
(Pdb)
> /Users/taugspurger/sandbox/fastparquet/fastparquet/dataframe.py(99)empty()
-> index = MultiIndex(levels=levels, labels=labels, names=index_names)
(Pdb) levels
[array(['1970-01-01T00:00:04.322432592', '1970-01-01T00:00:04.322432592',
       '1970-01-01T00:00:04.321907912'], dtype='datetime64[ns]'), array([None, None, None], dtype=object)]
 ```

Levels is supposed to be unique.

**Comment by 0x0L** [comment]
Actually I'm loosing the MultiIndex because I used the default constructor...
Using `from_arrays` looks better. 

The test is supposed to fail bc categories index are corrupted :)
It failed on your machine because of the init of the levels. Probably should use `check_integrity=False`

**Comment by 0x0L** [comment]
I can't reproduce the issue with the travis build in 3.5

File: fastparquet/dataframe.py [hunk 1]
```diff
@@ -65,39 +65,49 @@ def cat(col):
...
-        ]
-        labels = [
-            range(size) if str(t) != 'category' else views[n]
-            for t, n in zip(index_types, index_names)
-        ]
-        index = MultiIndex(levels=levels, labels=labels, names=index_names)
-    else:
-        index = indexes[0]
+                index._levels.append(Index(d))
+                index._labels.append(np.arange(size, dtype=int))
```
Comment (0x0L) [inline fastparquet/dataframe.py]: not too sure about the `np.arange(...)`

Comment (martindurant) [reply fastparquet/dataframe.py]: Yeah, I have no idea what goes in here.

**Comment by 0x0L** [comment]
The execution path has not changed if index is not a multi-index

**Comment by martindurant** [comment]
Any ideas on the remaining test failure?

**Comment by 0x0L** [comment]
I have no idea since I can't even reproduce it (in a conda env 3.5)
Since the test fails on `r.count` I don't even expect the code I changed to be run

**Comment by martindurant** [comment]
I kicked it and it passed - must have been a weird travis timing thing, which happens time to time.
Merging this now, thanks for the work, @0x0L

**Comment by pranav-kohli** [comment]
@martindurant , When will this commit come as a release version?

**Comment by martindurant** [comment]
I am at scipy this week, but I will try to fit a release in.

**Pull request merged**